// @vscode/wasm-wasi/lib/tools/main.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
/* eslint-disable no-console */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = exports.Options = void 0;
const paths = __importStar(require("node:path"));
const fs = __importStar(require("node:fs/promises"));
const yargs = __importStar(require("yargs"));
var Options;
(function (Options) {
    Options.defaults = {
        help: false,
        version: false,
        stdout: false,
        out: undefined,
        directory: undefined
    };
})(Options = exports.Options || (exports.Options = {}));
async function readDirectory(path, stat) {
    const name = paths.basename(path);
    if (!stat.isDirectory()) {
        throw new Error(`${path} is not a directory`);
    }
    const result = {
        kind: 'directory',
        name,
        size: stat.size,
        ctime: stat.ctimeNs,
        atime: stat.atimeNs,
        mtime: stat.mtimeNs,
        children: Object.create(null)
    };
    const dir = await fs.opendir(path);
    for await (const entry of dir) {
        const entryPath = paths.join(path, entry.name);
        const stat = await fs.stat(entryPath, { bigint: true });
        if (stat.isDirectory()) {
            result.children[entry.name] = await readDirectory(entryPath, stat);
        }
        else if (stat.isFile()) {
            result.children[entry.name] = {
                kind: 'file',
                size: stat.size,
                name: entry.name,
                ctime: stat.ctimeNs,
                atime: stat.atimeNs,
                mtime: stat.mtimeNs
            };
        }
        else {
            throw new Error(`${entryPath} is not a file or directory`);
        }
    }
    return result;
}
async function run(options) {
    if (options.help) {
        yargs.showHelp();
        return 0;
    }
    if (options.version) {
        console.log(require('../../package.json').version);
        return 0;
    }
    if (!options.directory) {
        console.error('Missing directory argument.');
        yargs.showHelp();
        return 1;
    }
    try {
        const stat = await fs.stat(options.directory, { bigint: true });
        if (!stat.isDirectory()) {
            console.error(`${options.directory} is not a directory`);
            return 1;
        }
        const directory = await readDirectory(options.directory, stat);
        directory.name = '/';
        const dump = JSON.stringify(directory, (_key, value) => typeof value === 'bigint' ? value.toString() : value, '\t');
        if (options.out) {
            await fs.writeFile(options.out, dump);
        }
        else if (options.stdout) {
            process.stdout.write(dump);
        }
        else {
            console.error('No output specified.');
            yargs.showHelp();
            return 1;
        }
        return 0;
    }
    catch (error) {
        console.error(`Creating dump failed`, error);
        return 1;
    }
}
async function main() {
    yargs.
        parserConfiguration({ 'camel-case-expansion': false }).
        exitProcess(false).
        usage(`Tool to generate a directory content for accessing in vscode.dev\nVersion: ${require('../../package.json').version}\nUsage: dir-dump [options] directory`).
        example(`dir-dump --stdout .`, `Create a directory dump for the current directory and print it to stdout.`).
        version(false).
        wrap(Math.min(100, yargs.terminalWidth())).
        option('v', {
        alias: 'version',
        description: 'Output the version number',
        boolean: true
    }).
        option('h', {
        alias: 'help',
        description: 'Output usage information',
        boolean: true
    }).
        option('out', {
        description: 'The output file the dump is save to.',
        string: true
    }).
        option('stdout', {
        description: 'Writes the dump to stdout.',
        boolean: true
    });
    const parsed = await yargs.argv;
    const options = Object.assign({}, Options.defaults, parsed);
    options.directory = parsed._[0];
    return run(options);
}
exports.main = main;
if (module === require.main) {
    main().then((exitCode) => process.exitCode = exitCode).catch((error) => { process.exitCode = 1; console.error(error); });
}
